windows/win_nssm.ps1

```diff
@@ -0,0 +1,547 @@
+#!powershell
+# This file is part of Ansible
+#
+# Copyright 2015, George Frank <george@georgefrank.net>
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+$ErrorActionPreference = "Stop"
+
+# WANT_JSON
+# POWERSHELL_COMMON
+
+$params = Parse-Args $args;
+$result = New-Object PSObject;
+Set-Attr $result "changed" $false;
+
+
+If ($params.name)
+{
+    $name = $params.name
+}
+Else
+{
+    Fail-Json $result "missing required argument: name"
+}
+
+If ($params.state)
+{
+    $state = $params.state.ToString().ToLower()
+    $validStates = "present", "absent", "started", "stopped", "restarted"
+    
+    # These don't really fit the declarative style of ansible
+    # If you need to do these things, you can just write a command for it
+    # "paused", "continued", "rotated"
+
+    If ($validStates -notcontains $state)
+    {
+        Fail-Json $result "state is $state; must be one of: $validStates"
+    }
+}
+else
+{
+    $state = "present"
+}
+
+If ($params.application)
+{
+    $application = $params.application
+}
+Else
+{
+    $application = $null
+}
+
+If ($params.app_parameters)
+{
+    $appParameters = $params.app_parameters
+}
+Else
+{
+    $appParameters = $null
+}
+
+If ($params.stdout_file)
+{
+    $stdoutFile = $params.stdout_file
+}
+Else
+{
+    $stdoutFile = $null
+}
+
+If ($params.stderr_file)
+{
+    $stderrFile = $params.stderr_file
+}
+Else
+{
+    $stderrFile = $null
+}
+
+Function Service-Exists
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    return ,[bool](Get-Service "$name" -ErrorAction SilentlyContinue)
+}
+
+Function Nssm-Remove
+{
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]$name
+    )
+
+    if (Service-Exists -name $name)
+    {
+        $cmd = "nssm stop ""$name"""
+        $results = invoke-expression $cmd
```

Comment:
The stop command can fail, yes; however, we're only interested in whether the service can be removed (since we're already sure it exists, per the `Service-Exists` check). Here are the scenarios that can play out:

1. The service is stopped. In this case, `nssm stop` will fail, but `nssm remove` will succeed, so we don't want to do anything about the `nssm stop` failure.
2. The service is hanging and won't respond. In this case, `nssm stop` will fail, and also `nssm remove` will fail, which is caught below.
3. The service is responding, but is not currently stopped. In this case, `nssm stop` will succeed, and so will `nssm remove`.

So this command is fine without an error check.